phpBB/phpbb/messenger/method/base.php

```diff
@@ -0,0 +1,518 @@
+ */
+abstract class base
+{
+	/** @var array */
+	protected $additional_headers = [];
+
+	/** @var assets_bag */
+	protected $assets_bag;
+
+	/** @var config */
+	protected $config;
+
+	/** @var dispatcher */
+	protected $dispatcher;
+
+	/** @var manager */
+	protected $ext_manager;
+
+	/** @var language */
+	protected $language;
+
+	/** @var log_interface */
+	protected $log;
+
+	/** @var string */
+	protected $msg = '';
+
+	/** @var queue */
+	protected $queue;
+
+	/** @var  path_helper */
+	protected $path_helper;
+
+	/** @var string */
+	protected $root_path;
+
+	/** @var  request */
+	protected $request;
+
+	/** @var string */
+	protected $subject = '';
+
+	/** @var \phpbb\template\template */
+	protected $template;
+
+	/** @var string */
+	protected $template_cache_path;
+
+	/** @var service_collection */
+	protected $twig_extensions_collection;
+
+	/** @var lexer */
+	protected $twig_lexer;
+
+	/** @var bool */
+	protected $use_queue = true;
+
+	/** @var user */
+	protected $user;
+
+	/**
+	 * Messenger base class constructor
+	 *
+	 * @param assets_bag $assets_bag
+	 * @param config $config
+	 * @param dispatcher $dispatcher
+	 * @param language $language
+	 * @param log_interface $log
+	 * @param request $request
+	 * @param user $user
+	 * @param queue $queue
+	 * @param path_helper $path_helper
+	 * @param manager $ext_manager
+	 * @param service_collection $twig_extensions_collection
+	 * @param lexer $twig_lexer
+	 * @param string $template_cache_path
+	 * @param string $phpbb_root_path
+	 */
+	function __construct(
+		assets_bag $assets_bag,
+		config $config,
+		dispatcher $dispatcher,
+		language $language,
+		log_interface $log,
+		request $request,
+		user $user,
+		queue $queue,
+		path_helper $path_helper,
+		manager $ext_manager,
+		service_collection $twig_extensions_collection,
+		lexer $twig_lexer,
+		$template_cache_path,
+		$phpbb_root_path
+	)
+	{
+		$this->assets_bag = $assets_bag;
+		$this->config = $config;
+		$this->dispatcher = $dispatcher;
+		$this->language = $language;
+		$this->log = $log;
+		$this->request = $request;
+		$this->user = $user;
+		$this->queue = $queue;
+		$this->path_helper = $path_helper;
+		$this->ext_manager = $ext_manager;
+		$this->twig_extensions_collection = $twig_extensions_collection;
+		$this->twig_lexer = $twig_lexer;
+		$this->template_cache_path = $template_cache_path;
+		$this->root_path = $phpbb_root_path;
+
+		$this->set_use_queue();
+	}
+
+	/**
+	 * Get messenger method id
+	 * @return mixed
+	 */
+	abstract public function get_id();
+
+	/**
+	 * Check if the messenger method is enabled
+	 * @return bool
+	 */
+	abstract public function is_enabled();
+
+	/**
+	 * Sets the use of messenger queue flag
+	 *
+	 * @return void
+	 */
+	public function set_use_queue($use_queue = true)
+	{
+		$this->use_queue = $use_queue;
+	}
+
+	/**
+	 * Resets all the data (address, template file, etc etc) to default
+	 *
+	 * @return void
+	 */
+	abstract public function reset();
```

Comment:
Maybe it makes sense to define an interface and only include base boilerplate code in the abstract class base?

Reply:
I don't know, I can do that but, to be honest, from my understanding having an interface while having at the same time an abstract class which implements only that one interface is pretty useless, as abstract class should implement all the interface declared methods anyway, so it won't make abstract class any lighter.
It would make sense if f.e. an [abstract] class would implement 2 or more different interfaces but this is not the case. So we'll end up in just having 1 class more (interface) for some reason.
May be just remove some `abstract` declarations which presumably shoudn't be mandatory for any messenger method?